File: java/src/main/java/io/cucumber/prettyformatter/LineBuilder.java
package io.cucumber.prettyformatter;

import io.cucumber.messages.types.TestStepResultStatus;
import io.cucumber.prettyformatter.Theme.Element;

import java.util.function.Consumer;

import static java.lang.System.lineSeparator;

final class LineBuilder {

    private final StringBuilder builder = new StringBuilder(80);
    private final Theme theme;
    private int unstyledLength;

    LineBuilder(Theme theme) {
        this.theme = theme;
    }

    LineBuilder indent(int indent) {
        return append(createPadding(indent));
    }

    LineBuilder title(Element keywordElement, String keyword, Element nameElement, String name) {
        return append(keywordElement, keyword + ":")
                .append(" ")
                .append(nameElement, name);
    }

    LineBuilder addPaddingUpTo(int index) {
        return append(createPadding(index));
    }

    private String createPadding(int commentStartAtIndex) {
        int padding = commentStartAtIndex - unstyledLength;
        StringBuilder builder = new StringBuilder(padding);
        for (int i = 0; i < padding; i++) {
            builder.append(" ");
        }
        return builder.toString();
    }

    LineBuilder accept(Consumer<LineBuilder> consumer) {
        consumer.accept(this);
        return this;
    }

    LineBuilder newLine() {
        unstyledLength = 0;
        builder.append(lineSeparator());
        return this;
    }

    LineBuilder append(String text) {
        this.unstyledLength += text.length();
        builder.append(text);
        return this;
    }

    LineBuilder append(Element element, String text) {
        this.unstyledLength += text.length();
        builder.append(theme.style(element, text));
        return this;
    }

    LineBuilder begin(Element element) {
        builder.append(theme.beginStyle(element));
        return this;
    }

    LineBuilder begin(Element element, TestStepResultStatus status) {
        builder.append(theme.beginStyle(element, status));
        return this;
    }

    LineBuilder end(Element element) {
        builder.append(theme.endStyle(element));
        return this;
    }

    LineBuilder end(Element element, TestStepResultStatus status) {
        builder.append(theme.endStyle(element, status));
        return this;
    }

    String build() {
        return builder.toString();
    }

    @Override
    public String toString() {
        return builder.toString();
    }

}

File: cpp/src/cucumber/pretty-formatter/LineBuilder.cpp
#include "cucumber/pretty-formatter/LineBuilder.hpp"
#include "cucumber/messages/TestStepResultStatus.hpp"
#include "cucumber/pretty-formatter/Theme.hpp"
#include <cstddef>
#include <memory>
#include <string>
#include <string_view>
#include <utility>

namespace cucumber::pretty_formatter
{
    LineBuilder::LineBuilder(std::shared_ptr<Theme> theme)
        : theme{ std::move(theme) }
    {
        string.reserve(initialCapacity);
    }

    LineBuilder& LineBuilder::Indent(std::size_t indent)
    {
        return Append(CreatePadding(indent));
    }

    LineBuilder& LineBuilder::AddPaddingUpTo(std::size_t indent)
    {
        return Append(CreatePadding(indent));
    }

    LineBuilder& LineBuilder::Title(Theme::Element element, const std::string& keyword, Theme::Element nameElement, const std::string& name)
    {
        return Append(element, keyword + ":").Append(" ").Append(nameElement, name);
    }

    LineBuilder& LineBuilder::StatusIcon(const std::string& icon)
    {
        unstyledLength += Theme::statusIconLength;
        string.append(icon);
        return *this;
    }

    LineBuilder& LineBuilder::NewLine()
    {
        unstyledLength = 0;
        string.append("\n");
        return *this;
    }

    LineBuilder& LineBuilder::Append(std::string_view text)
    {
        unstyledLength += text.size();
        string.append(text);
        return *this;
    }

    LineBuilder& LineBuilder::Append(Theme::Element element, std::string_view text)
    {
        unstyledLength += text.size();
        string.append(theme->Style(element, text));
        return *this;
    }

    LineBuilder& LineBuilder::Append(Theme::Element element, messages::TestStepResultStatus status, std::string_view text)
    {
        unstyledLength += text.size();
        string.append(theme->Style(element, status, text));
        return *this;
    }

    LineBuilder& LineBuilder::Begin(Theme::Element element)
    {
        string.append(theme->BeginStyle(element));
        return *this;
    }

    LineBuilder& LineBuilder::Begin(Theme::Element element, messages::TestStepResultStatus status)
    {
        string.append(theme->BeginStyle(element, status));
        return *this;
    }

    LineBuilder& LineBuilder::End(Theme::Element element)
    {
        string.append(theme->EndStyle(element));
        return *this;
    }

    LineBuilder& LineBuilder::End(Theme::Element element, messages::TestStepResultStatus status)
    {
        string.append(theme->EndStyle(element, status));
        return *this;
    }

    std::string LineBuilder::Build()
    {
        string.resize(string.size());
        return std::move(string);
    }

    std::string LineBuilder::CreatePadding(std::size_t indent) const
    {
        if (indent <= unstyledLength)
        {
            return std::string{};
        }

        const auto padding = indent - unstyledLength;
        return std::string(padding,
            ' '); // NOLINT(modernize-return-braced-init-list): explicit constructor can't be called using braced initializers.
    }
}
